compiler/runtime/OMRRelocationRecord.hpp

```diff
@@ -0,0 +1,212 @@
+
+#ifndef OMR_RELOCATION_RECORD_INCL
+#define OMR_RELOCATION_RECORD_INCL
+/*
+ * The following #define(s) and typedef(s) must appear before any #includes in this file
+ */
+#ifndef OMR_RELOCATION_RECORD_CONNECTOR
+#define OMR_RELOCATION_RECORD_CONNECTOR
+namespace OMR { class RelocationRecord; }
+namespace OMR { typedef OMR::RelocationRecord RelocationRecordConnector; }
+#else
+   #error OMR::RelocationRecord expected to be a primary connector, but another connector is already defined
+#endif
+
+
+#ifndef OMR_RELOCATION_RECORD_BINARY_TEMPLATE_CONNECTOR
+#define OMR_RELOCATION_RECORD_BINARY_TEMPLATE_CONNECTOR
+namespace OMR { class RelocationRecordBinaryTemplate; }
+namespace OMR { typedef OMR::RelocationRecordBinaryTemplate RelocationRecordBinaryTemplateConnector; }
+#else
+   #error OMR::RelocationRecordBinaryTemplate expected to be a primary connector, but another connector is already defined
+#endif
+
+
+#include <stdint.h>
+#include "compile/Compilation.hpp"
+#include "env/jittypes.h"
+#include "infra/Link.hpp"
+#include "infra/Flags.hpp"
+#include "runtime/TRRelocationRuntime.hpp"
+#include "runtime/Runtime.hpp"
+
+namespace TR {
+class RelocationRuntime;
+class RelocationTarget;
+class RelocationRecordBinaryTemplate;
+struct RelocationRecordMethodCallPrivateData
+   {
+   uintptr_t callTargetOffset;
+   };
+
+union RelocationRecordPrivateData
+   {
+   int storage;
+   };
+
+}
+
+namespace OMR { typedef TR_ExternalRelocationTargetKind RelocationRecordType;}
+namespace TR { typedef TR_ExternalRelocationTargetKind RelocationRecordType;}
+namespace TR { class RelocationRecordBinaryTemplate;}
+extern char* AOTcgDiagOn;
+
+// TR::RelocationRecord is the base class for all relocation records.  It is used for all queries on relocation
+// records as well as holding all the "wrapper" parts.  These classes are an interface to the *BinaryTemplate
+// classes which are simply structs that can be used to directly access the binary representation of the relocation
+// records stored in the cache (*BinaryTemplate structs are defined near the end of this file after the
+// RelocationRecord* classes.  The RelocationRecord* classes permit virtual function calls access to the
+// *BinaryTemplate classes and must access the binary structs via the _record field in the TR::RelocationRecord
+// class.  Most consumers should directly manipulate the TR::RelocationRecord* classes since they offer
+// the most flexibility.
+// No class that derives from TR::RelocationRecord should define any state: all state variables should be declared
+//  in TR::RelocationRecord or the constructor/decode() mechanisms will not work properly
+// Relocation record classes for "real" relocation record types
+// should be defined on the level of consuming project
+
+namespace OMR
+{
+
+class OMR_EXTENSIBLE RelocationRecordBinaryTemplate
+   {
+public:
+   RelocationRecordBinaryTemplate(){};
+   TR::RelocationRecordBinaryTemplate* self();
```

Comment:
The `self()` function should be `protected`, not `public`.

Reply:
Thank you for the comment! Will address all of them!